### PR TITLE
- classes human/employee.

### PR DESCRIPTION
**Pull request by AndrewQA88**
- calculate age, job experience, age when got a job.

### CODE REVIEW

File: src/test/java/humanemployeeexercise/Employee.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        super(birthDate);
+        this.dateOfEmployment = dateOfEmployment;
+    }
+
+    public ZonedDateTime getDateOfEmployment() {
+        return dateOfEmployment;
+    }
+
+    public long ageOfEmployee(ZonedDateTime birthDate, ZonedDateTime currentDate) {
+        return ChronoUnit.YEARS.between(birthDate, currentDate);
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Employee.java:20]: currentDate.getYear-birthDate.getYear
check is current Month is  after birth month
and the same for date

File: src/test/java/humanemployeeexercise/Employee.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    public long ageWhenStartedWorking(ZonedDateTime birthDate, ZonedDateTime hiringDate) {
+        return ChronoUnit.YEARS.between(birthDate, hiringDate);
+    }
+
+    public long jobExperience(ZonedDateTime hiringDate, ZonedDateTime currentDate) {
+        return ChronoUnit.YEARS.between(hiringDate, currentDate);
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Employee.java:28]: is difference if you ask job expirirence that was on some date or will be in some future date and for now
for now to  check current expirince jut use here directly ZonedDateTime.now()

File: src/test/java/humanemployeeexercise/Employee.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    Employee(ZonedDateTime birthDate, ZonedDateTime dateOfEmployment) {
+        super(birthDate);
+        this.dateOfEmployment = dateOfEmployment;
+    }
+
+    public ZonedDateTime getDateOfEmployment() {
+        return dateOfEmployment;
+    }
+
+    public long ageOfEmployee(ZonedDateTime birthDate, ZonedDateTime currentDate) {
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Employee.java:19]: getAgeOfEmployee with out parameters

**Review by iikarak — CHANGES_REQUESTED** [review]

File: src/test/java/humanemployeeexercise/Employee.java (hunk 2 above)
Comment (AndrewQA88) [reply src/test/java/humanemployeeexercise/Employee.java:28]: changed directly to ZonedDateTime.now()

File: src/test/java/humanemployeeexercise/Employee.java (hunk 3 above)
Comment (AndrewQA88) [reply src/test/java/humanemployeeexercise/Employee.java:19]: replace parameters with getAgeOfEmployee and getDateOfEmployment() directly in method

File: src/test/java/humanemployeeexercise/Employee.java (hunk 1 above)
Comment (AndrewQA88) [reply src/test/java/humanemployeeexercise/Employee.java:20]: public long jobExperience() {
        if (ZonedDateTime.now().getMonth().getValue() == getDateOfEmployment().getMonth().getValue()
                && ZonedDateTime.now().getDayOfMonth() < getDateOfEmployment().getDayOfMonth()) {
            return (ZonedDateTime.now().getYear() - getDateOfEmployment().getYear() - 1);
        } else if (ZonedDateTime.now().getMonth().getValue() < getDateOfEmployment().getMonth().getValue()) {
            return (ZonedDateTime.now().getYear() - getDateOfEmployment().getYear() - 1);
        } else {
            return ZonedDateTime.now().getYear() - getDateOfEmployment().getYear();
        }
    }

File: src/test/java/humanemployeeexercise/Employee.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        this.dateOfEmployment = dateOfEmployment;
+        this.title = title;
+    }
+
+    public ZonedDateTime getDateOfEmployment() {
+        return dateOfEmployment;
+    }
+
+    public long ageOfEmployee() {
+        long fullYears = ZonedDateTime.now().getYear() - getBirthDate().getYear();
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Employee.java:28]: for counting of amount of full years based on date please create one function and use it in all places where it is requred, in order to not have code duplictaion

File: src/test/java/humanemployeeexercise/Employee.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                && ZonedDateTime.now().getDayOfMonth() < getBirthDate().getDayOfMonth())) {
+            return fullYears - 1;
+        } else {
+            return fullYears;
+        }
+    }
+
+    public long ageWhenStartedWorking() {
+        long fullYears = getDateOfEmployment().getYear() - getBirthDate().getYear();
+        if ((getDateOfEmployment().getMonth().getValue() < getBirthDate().getMonth().getValue())
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Employee.java:40]: please review comparison operators that we have in java to form smaller condition inside if check statement

File: src/test/java/humanemployeeexercise/Human.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
+package humanemployeeexercise;
+
+import java.time.ZonedDateTime;
+
+public class Human implements Creature {
+
+    private final ZonedDateTime birthDate;
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Human.java:7]: good usage of final

File: src/test/java/humanemployeeexercise/Creature.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package humanemployeeexercise;
+
+public interface Creature {
+
+    int getWeight();
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Creature.java:5]: add default function into interface as was requeted in task description

**Review by iikarak — CHANGES_REQUESTED** [review]

File: src/test/java/humanemployeeexercise/Creature.java (hunk 7 above)
Comment (AndrewQA88) [reply src/test/java/humanemployeeexercise/Creature.java:5]: default void getMeaning(){
        System.out.println("Creature: something created either animate or inanimate.");
    }

File: src/test/java/humanemployeeexercise/Employee.java (hunk 5 above)
Comment (AndrewQA88) [reply src/test/java/humanemployeeexercise/Employee.java:40]: done

File: src/test/java/humanemployeeexercise/Employee.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                || (firstDate.getMonth().getValue() == secondDate.getMonth().getValue() && firstDate.getDayOfMonth() < secondDate.getDayOfMonth())) {
+            return fullYears - 1;
+        } else {
+            return fullYears;
+        }
+    }
+
+    public long ageOfEmployee() {
+        ZonedDateTime currentDate = ZonedDateTime.now();
+        return calculateFullYear(currentDate, getBirthDate());
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Employee.java:39]: use just variable of class with out get

File: src/test/java/humanemployeeexercise/Employee.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    public long ageWhenStartedWorking() {
+        return calculateFullYear(getDateOfEmployment(), getBirthDate());
+    }
+
+    public long jobExperience() {
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Employee.java:46]: rename functions who have return type not void

File: src/test/java/humanemployeeexercise/Human.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    public ZonedDateTime getBirthDate() {
+        return birthDate;
+    }
+
+    public String getName() {
+        return name;
+    }
+
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Human.java:31]: @Override

File: src/test/java/humanemployeeexercise/Human.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    public String toString() {
+        return this.GREETING + "\n" + "Name: " + this.name;
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Human.java:33]: add print whole info

File: src/test/java/humanemployeeexercise/Employee.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    public int getSalary() {
+        return salary;
+    }
+
+    public void setSalary(int salary) {
+        this.salary = salary;
+    }
+
+    public String toString() {
+        return this.GREETING + "\n" + "Name: " + super.getName() + "\n" + "Title: " + this.title;
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Employee.java:60]: use method toString from parent class

File: src/test/java/humanemployeeexercise/Employee.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    public void setVacation(int year, int month) {
+        this.vacation = ZonedDateTime.of(LocalDate.of(year, month, 1), LocalTime.of(0, 0), ZoneId.of("Europe/Helsinki"));
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Employee.java:64]: to not duplicate code, reuse here function with 3 parameters

**Review by iikarak — CHANGES_REQUESTED** [review]

File: src/test/java/humanemployeeexercise/Employee.java (hunk 8 above)
Comment (AndrewQA88) [reply src/test/java/humanemployeeexercise/Employee.java:39]: updated with getDateOfEmployment(), birthDate is the superclass's field

File: src/test/java/humanemployeeexercise/Employee.java (hunk 9 above)
Comment (AndrewQA88) [reply src/test/java/humanemployeeexercise/Employee.java:46]: done

File: src/test/java/humanemployeeexercise/Human.java (hunk 10 above)
Comment (AndrewQA88) [reply src/test/java/humanemployeeexercise/Human.java:31]: done

File: src/test/java/humanemployeeexercise/Human.java (hunk 11 above)
Comment (AndrewQA88) [reply src/test/java/humanemployeeexercise/Human.java:33]: done

File: src/test/java/humanemployeeexercise/Employee.java (hunk 12 above)
Comment (AndrewQA88) [reply src/test/java/humanemployeeexercise/Employee.java:60]: done

File: src/test/java/humanemployeeexercise/Employee.java (hunk 13 above)
Comment (AndrewQA88) [reply src/test/java/humanemployeeexercise/Employee.java:64]: public void setVacation(int year, int month) {
        setVacation(year, month, 1);
    }

File: src/test/java/humanemployeeexercise/Company.java [hunk 14]
```diff
@@ -0,0 +1,17 @@
+package humanemployeeexercise;
+
+public class Company {
+
+    public void allowance(Employee employee) {
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Company.java:5]: try with update of int salary

Comment (iikarak) [reply src/test/java/humanemployeeexercise/Company.java:5]: allowance(int salary) {
salary+=500;
} 
salary?

**Review by iikarak — CHANGES_REQUESTED** [review]

File: src/test/java/humanemployeeexercise/Employee.java (hunk 4 above)
Comment (AndrewQA88) [reply src/test/java/humanemployeeexercise/Employee.java]: done

File: src/test/java/humanemployeeexercise/Company.java (hunk 14 above)
Comment (iikarak) [reply src/test/java/humanemployeeexercise/Company.java]: new Company().allowance(andriy.getSalary());
1-> allowance void -> no incrise
2-> allowance return int -> andriy.setSalary(new Company().allowance(andriy.getSalary()))

File: src/test/java/humanemployeeexercise/Director.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+public class Director extends Employee {
+
+    int salary;
+
+    public int getSalary(Object requester) {
+        if (requester instanceof Human) {
+            return salary;
+        } else {
+            try {
+                throw new IncorrectAccessException("access denied");
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Director.java:17]: with out try chatch

File: src/test/java/humanemployeeexercise/Director.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+            } catch (IncorrectAccessException e) {
+                e.printStackTrace();
+            }
+        }
+        return -1;
```
Comment (iikarak) [inline src/test/java/humanemployeeexercise/Director.java:22]: INDEX_NOT_FOUND
org.apache.commons.lang3.StringUtils

**Review by iikarak — APPROVED** [review]

**Pull request merged**